protocol/messenger_unread_test.go

```diff
@@ -0,0 +1,277 @@
+	chat := CreatePublicChat("test-chat-1", s.m.transport)
+
+	chat.UnviewedMessagesCount = 2
+	chat.UnviewedMentionsCount = 2
+	chat.Highlight = true
+
+	err := s.m.SaveChat(chat)
+	s.Require().NoError(err)
+
+	inputMessage1 := buildTestMessage(*chat)
+	inputMessage1.ID = "1"
+	inputMessage1.Seen = true
+	inputMessage1.Mentioned = true
+
+	err = s.m.SaveMessages([]*common.Message{inputMessage1})
+	s.Require().NoError(err)
+
+	_, err = s.m.MarkAllRead(context.Background(), chat.ID)
+	s.Require().NoError(err)
+
+	chats := s.m.allChats
+
+	actualChat, ok := chats.Load(chat.ID)
+
+	s.Require().Equal(true, ok)
+	s.Require().Equal(uint(0), actualChat.UnviewedMessagesCount)
+	s.Require().Equal(uint(0), actualChat.UnviewedMentionsCount)
+	s.Require().Equal(false, actualChat.Highlight)
+
+	actualMessage, err := s.m.MessageByID(inputMessage1.ID)
+	s.Require().NoError(err)
+	s.Require().Equal(true, actualMessage.Seen)
+
+	count, countWithMentions, err := s.m.MarkMessageAsUnread(chat.ID, inputMessage1.ID)
+	s.Require().NoError(err)
+	s.Require().Equal(uint64(1), count)
+	s.Require().Equal(uint64(1), countWithMentions)
+
+	chats = s.m.allChats
+
+	actualChat, ok = chats.Load(chat.ID)
+	s.Require().Equal(true, ok)
+	s.Require().Equal(uint(1), actualChat.UnviewedMessagesCount)
+	s.Require().Equal(uint(1), actualChat.UnviewedMentionsCount)
+	s.Require().Equal(false, actualChat.Highlight)
+
+	actualMessage, err = s.m.MessageByID(inputMessage1.ID)
+	s.Require().NoError(err)
+	s.Require().Equal(false, actualMessage.Seen)
+}
+
+func (s *MessengerSuite) TestMarkMessageAsUnreadWhenMessageListContainsSeveralMessages() {
+	chat := CreatePublicChat("test-chat-2", s.m.transport)
+
+	chat.UnviewedMessagesCount = 0
+	chat.UnviewedMentionsCount = 0
+	chat.Highlight = true
+
+	err := s.m.SaveChat(chat)
+	s.Require().NoError(err)
+
+	inputMessage1 := buildTestMessage(*chat)
+	inputMessage1.ID = "1"
+	inputMessage1.Seen = true
+	inputMessage1.Mentioned = true
+
+	time.Sleep(1000 * time.Millisecond)
+
+	inputMessage2 := buildTestMessage(*chat)
+	inputMessage2.ID = "2"
+	inputMessage2.Seen = true
+	inputMessage2.Mentioned = false
+
+	time.Sleep(1000 * time.Millisecond)
+
+	inputMessage3 := buildTestMessage(*chat)
+	inputMessage3.ID = "3"
+	inputMessage3.Seen = true
+	inputMessage3.Mentioned = false
+
+	err = s.m.SaveMessages([]*common.Message{
+		inputMessage1,
+		inputMessage2,
+		inputMessage3,
+	})
+	s.Require().NoError(err)
+
+	_, err = s.m.MarkAllRead(context.Background(), chat.ID)
+	s.Require().NoError(err)
+
+	chats := s.m.allChats
+	actualChat, ok := chats.Load(chat.ID)
+
+	s.Require().Equal(true, ok)
+	s.Require().Equal(uint(0), actualChat.UnviewedMessagesCount)
+	s.Require().Equal(uint(0), actualChat.UnviewedMentionsCount)
+	s.Require().Equal(false, actualChat.Highlight)
+
+	actualMessage1, err := s.m.MessageByID(inputMessage1.ID)
+	s.Require().NoError(err)
+	s.Require().Equal(true, actualMessage1.Seen)
```

Comment:
So you could make a separate function and simplify the code a lot.

```go
func (s *MessengerSuite) checkMessageSeen(messageID string, expectedSeen) {
    message, err := s.m.MessageByID(messageID)
	s.Require().NoError(err)
	s.Require().Equal(expectedSeen, message.Seen)
}

s.checkMessageSeen(inputMessage1.ID, true)
s.checkMessageSeen(inputMessage2.ID, true)
s.checkMessageSeen(inputMessage3.ID, true)
```

Reply:
Sure, updated. thanks